hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -1942,30 +1967,31 @@ def reoffset_pair(pair, o):
             m = min([int_from_block(p) for p in valid_pair])
 
             new_blocks = [self.shrink_target.buffer[u:v]
-                          for u, v in self.blocks]
+                          for u, v in self.shrink_target.each_block_bounds()]
             for i in valid_pair:
                 new_blocks[i] = int_to_bytes(
                     int_from_block(i) + o - m, block_len(i))
             buffer = hbytes().join(new_blocks)
             return self.incorporate_new_buffer(buffer)
 
-        i = 0
-        while i < len(self.blocks):
-            if self.is_payload_block(i) and int_from_block(i) > 0:
-                j = i + 1
-                while j < len(self.blocks):
-                    block_val = int_from_block(j)
-                    i_block_val = int_from_block(i)
-                    if self.is_payload_block(j) \
-                       and block_val > 0 and i_block_val > 0:
-                        offset = min(int_from_block(i),
-                                     int_from_block(j))
-                        Integer.shrink(
-                            offset, lambda o: reoffset_pair((i, j), o),
-                            random=self.random
-                        )
-                    j += 1
-            i += 1
+        def is_non_zero_payload(block):
+            return not block.all_zero and self.is_payload_block(block.index)
+
+        for block_i, block_j in self.each_pair_of_blocks(
+            is_non_zero_payload,
+            is_non_zero_payload,
+        ):
+            i = block_i.index
+            j = block_j.index
+
+            value_i = int_from_block(i)
```

Comment:
This is not a request for change, more of a thought for discussion: When I was looking at the block class, I was thinking it was a shame that you couldn't access its contents from the class. I wonder if it would make sense to give `Block` a weak reference to the enclosing data so that that could be calculated on demand (actually storing it for each block would be quite memory hungry), and then value this be `value_i = block_i.as_int()` or something.

Reply:
I had been thinking about similar helper methods, for slicing block data out of a buffer and for splicing it back into a candidate buffer.

(One thing to be careful of is keeping track of which buffer we actually want to read from. This can be subtle when running one of the `shrinking/` kernels, and potentially updating the shrink target between iterations.)